internal/pkg/debug/graph/graph.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package graph defines an abstraction of the SSA graph that facilitates rendering.
+package graph
+
+import (
+	"errors"
+
+	"golang.org/x/tools/go/ssa"
+)
+
+type relationship int
+
+const (
+	// Referrer represents a node that is referred to as per ssa.Value.Referrers
+	Referrer relationship = iota
+	// Operand represents a node that is an operand of an instruction as per ssa.Instr.Operands
+	Operand
+)
+
+// Node represents a node in the SSA graph, along with its relationship to a parent node.
+type Node struct {
+	// N is the ssa.Node wrapped by this type.
+	N ssa.Node
+	// R is the relationship between this node and its parent in the SSA graph.
+	R relationship
+}
+
+// FuncGraph represents the SSA graph for an ssa.Function.
+type FuncGraph struct {
+	// F is the function whose graph this is
+	F *ssa.Function
+	// Children is a mapping from each node to its children (referrers + operands)
+	Children map[ssa.Node][]Node
+	// visited is used while creating the graph to avoid needlessly revisiting nodes
+	visited map[ssa.Node]bool
+	// stack is used to perform a DFS on F's SSA graph
+	stack
+}
+
+// New returns a new Graph constructed from a given function.
+func New(f *ssa.Function) *FuncGraph {
+	g := FuncGraph{
+		F:        f,
+		Children: map[ssa.Node][]Node{},
+		visited:  map[ssa.Node]bool{},
+	}
+	g.visitBlocks()
+	return &g
+}
+
+func (g *FuncGraph) visitBlocks() {
+	for _, b := range g.F.Blocks {
+		g.visit(b)
+	}
+}
+
+func (g *FuncGraph) visit(b *ssa.BasicBlock) {
+	if len(b.Instrs) == 0 {
+		return
+	}
+
+	n := b.Instrs[0].(ssa.Node)
+	g.visited[n] = true
+	g.push(n)
+	for len(g.stack) > 0 {
+		current, err := g.pop()
+		if err != nil {
+			break
+		}
```

Comment:
A failure to pop would indicate a deeper problem.  It warrants at least logging, if not fully propagating up to end execution.  Alternatively, if we're not consuming the error, there's no reason to emit one below.

Reply:
I've added logging for now. A `pop` on an empty stack would almost certainly mean an error in the graph traversal code, which we would be the most likely ones to catch while running `go test -debug`. In that light, I think logging could be enough, actually.

Reply:
Works for me.